### PR TITLE
Sentry and error handling refactor

### PR DESCRIPTION
**Pull request by stristr**
OK to merge, but I recommend testing!

Also might want some design love from @taylorpoe, so I'm requesting his review to try out the UX with instructions below.

Medium review.

Summary of changes:

- Fixes [Sentry isn't getting some crashes](https://app.asana.com/0/824697491061885/803720498732763/f)
- Fixes [If carbonite upload fails or hangs, Haiku never closes on crash](https://app.asana.com/0/506410768732347/792533506168501/f)
- Moves Carbonite (DAL) and error reporting (BLL) work to Envoy.
- Adds ability to test Sentry in dev via an environment variable. To test in dev, you can:
   - Add `DEBUG_SENTRY=1` to `~/.haiku/.env`
   - Restart the app.
   - Force errors, either by triggering a known existing bug or e.g. by putting `setTimeout(() => { throw new Error('whoospy!') }, 5000)` in the `constructor` function in `Creator.js`. Please be sure to test that errors work in Master (Plumbing), Timeline, Glass, _and_ Creator.
- Wires up the fail whale and replaces all known ways to insta-crash with the fail whale.

Regressions to look for:

- Anything related to crashes and error handling, both in dev and prod.

Completed checkin tasks:

- [x] Did manual testing of interrelated functionality
- [x] Updated `changelog/public/latest.json`

### CODE REVIEW

**Review by tsortwehttam — COMMENTED** [review]
I'm in the midst of reviewing this but blocked currently due to a compilation problem with (at least) these two modules:

```
packages/haiku-glass/src/react/index.js
packages/haiku-timeline/src/index.js
```

The following line, which I would expect to see in the compiled output files —

```
var ProjectDefinitions_1 = require("@haiku/sdk-client/lib/ProjectDefinitions");
```

— seems to be missing from the compiled output files. This resulting in a `ReferenceError: ProjectDefinitions_1 is not defined` crash when trying to open any project.

This does not appear to be an issue in `master`.

**Comment by stristr** [comment]
So…does `packages/@haiku/sdk-client/src/ProjectDefinitions.ts` not exist for you?

And did you try `yarn compile-all --force`?

**Comment by stristr** [comment]
@matthewtoast, figured out what this was (it was me, not you).

Ready for a re-check.

**Comment by tsortwehttam** [comment]
Code LGTM and locally reviewed. No question this is working better than before.

But I wanted to note a few edge-case/races that are _probably_ expected:

- In Creator.js constructor:
  - `setTimeout(() => {throw new Error('TEST')}, 16);` ~> Sentry error; no failwhale.
  - `throw new Error('TEST')` ~> Sentry error; app stops at gray screen.
- In Timeline.js constructor:
  - `setTimeout(() => {throw new Error('TEST 2')}, 16);` ~> Sentry error; no failwhale.
  - `throw new Error('TEST')` ~> Sentry error; app stops at "H" loader.
  - `throw new Error('TEST 3');` in `Project.setup` callback ~> Sentry error; app stops at "H" loader.
- In Glass.js constructor:
  - `setTimeout(() => {throw new Error('TEST 4')}, 16);` ~> No error; no failwhale, no toast.
- Elsewhere:
  - ActiveComponent.js#L1319 `return done(new Error('TEST 6'));` ~> Toast, but no Sentry.

tl;dr...

- If uncaught exceptions happen to soon, no failwhale.
- It seems like _returned errors from ActiveComponent callbacks_ don't go to Sentry.

Finally, if we want to go full commercial airline, we can make the changelog say:

> In the unlikely event of a water landing

😛

**Comment by stristr** [comment]
Thanks for the review. The situation where crashing before Envoy channels can be resolved prohibits correct error handling is sort of a feature, or at least it's known and intentional—I figured we can take at least some things for granted, like "IPC works and is a suitable conduit for our custom crash handling behavior when a crash occurs."

Re: `ActiveComponent.js`, just to clarify—you see a toast, but no fail whale (and therefore no crash)?

I thought that was handled [here](https://github.com/HaikuTeam/mono/pull/845/files#diff-50d5a3593264a0972064026fd17c543dR549), but I will take another look before merging.

**Comment by stristr** [comment]
I got pretty far with a pretty good solution for also sending ActiveComponent errors to Sentry (tl;dr, send _all_ `logger.error(...)` things to Sentry), but hit some snags that made it not totally straightforward. I'm going to merge this as is and mark an item for later in mono2.

File: packages/haiku-creator/src/react/components/library/Library.js [hunk 1]
```diff
@@ -212,7 +212,6 @@ class Library extends React.Component {
         const message = error.err || 'We had a problem connecting with Figma. Please check your internet connection and try again.';
         const reportData = {url, message};
 
-        Raven.captureException(new Error(message), reportData);
```
Comment (stristr) [inline packages/haiku-creator/src/react/components/library/Library.js:215]: Heads up @roperzh—`Raven.captureException` now triggers the fail whale, so I had to rip this out. Stay tuned for a better way to do this—I figured the data's in mixpanel and the "exception" always happens in the same place, so we probably don't need carbonite/Sentry for this….
